ide/coqide/coqOps.ml

```diff
@@ -146,6 +138,54 @@ end = struct
 end
 open SentenceId
 
+(* Given a Coq loc, convert it to a pair of iterators start / end in
+   the buffer. *)
+let coq_loc_to_gtk_offset (buffer : GText.buffer) loc =
+  buffer#get_iter_at_byte ~line:loc.Loc.line_nb (loc.bp - loc.bol_pos),
+  buffer#get_iter_at_byte ~line:loc.Loc.line_nb_last (loc.ep - loc.bol_pos_last)
+
+(** increase [uni_off] by the number of bytes until [s_uni] This can
+    be used to convert a character offset to byte offset if we know a
+    previous point denoted by the pair [s_byte, s_uni] *)
+let c2b (buffer : GText.buffer) (s_byte, s_uni) uni_off =
+  if s_uni < 0 then
+    raise (invalid_arg "c2b: s_uni must be >= 0");
+  let rec cvt iter rv =
+    if iter#offset <= s_uni then
+      rv
+    else
+      cvt iter#backward_char (rv + (Ideutils.ulen iter#char))
+  in
+  cvt (buffer#get_iter (`OFFSET uni_off)) s_byte
+
+(* Given [sentence], returns [bp, line, bol] where [bp] is the
+   absolute offset in _bytes_ for the start of [sentence], [line] is
+   the line number for the start of sentence and [bol] is the line
+   ofsset of [sentence] in _bytes_.
+
+   This is meant to be used in the resumption of parsing, to tell Coq
+   where the input stream was left at. In this case, these 3 points
+   uniquely determine the position where to restart
+   from. [CLexer.after] can compute this from the position of the last
+   parsed token.
+
+   However, for CoqIDE, we have a huge problem in the sense that
+   CoqIDE will do its own parsing and the locations may not start at
+   the same point. Thus, we need to perform potentially very costly
+   char to offset conversion, for that we need the whole buffer.
```

Comment:
Would you elaborate on this?  What do you mean by "CoqIDE will do its own parsing and locations may not start at the same point?"

Reply:
CoqIde uses `coq_lex.mll` to split sentences, whereas Coq uses it's own parser, nothing ensures they will pick the same exact splitting point, due to comments, etc...

Reply:
CoqIDE may split sentences a bit differently from another IDE, but I don't see how that's a huge problem.  Presumably `get_iter_at_byte` does the conversion efficiently enough that there's no user-visible perf issue.

One thing we could do more neatly is to always return `loc` with errors and warnings so we use the parser's notion of sentence boundaries.  Currently we highlight code based on CoqIDE's notion, so the highlight includes leading white space and comments.  Making this change would require addressing  the use case in #17023; notice that the first screen shot there doesn't highlight the white space.

![image](https://user-images.githubusercontent.com/1253341/209579520-e80acadc-da84-4a12-a5e8-dc798bf37abe.png)

Reply:
> CoqIDE may split sentences a bit differently from another IDE, but I don't see how that's a huge problem. Presumably get_iter_at_byte does the conversion efficiently enough that there's no user-visible perf issue.

Indeed, `get_iter_at_byte` is the direction that is not problematic; in the other direction we can't rely on Coq's loc as to avoid scanning the whole buffer. The problem is not CoqIDE splitting sentences differently from other IDEs, but from Coq itself, and thus we can't reuse the line work.

> One thing we could do more neatly is to always return loc with errors and warnings so we use the parser's notion of sentence boundaries.

Sane document managers already do that. However often errors are more specific than a sentence.

> Currently we highlight code based on CoqIDE's notion, so the highlight includes leading white space and comments. Making this change would require addressing the use case in https://github.com/coq/coq/issues/17023; notice that the first screen shot there doesn't highlight the white space.

that's a problem with CoqIDE, coq side should be mostly good. The above example works fine for example in `coq-lsp`.